numpy/_core/src/umath/string_ufuncs.cpp

```diff
@@ -777,6 +893,36 @@ init_string_ufuncs(PyObject *umath)
         return -1;
     }
 
+    dtypes[0] = dtypes[2] = NPY_OBJECT;
+    dtypes[1] = NPY_INT64;
```

Comment:
Oh this is a great point. I fixed the type resolution and added tests for different integer dtypes. Also, how did you manage to do a promoter for a ufunc that also has the old type resolution thing? I remember having difficulties with that when implementing `np.add`.

Reply:
I think you might have to convert over to the new arraymethod API to use a promoter (although @seberg would know for sure). It would also work to e.g. template the loop and add loops for each integer type, but that hurts code size and is also tricky because there are unintuitive rules about which integer types are aliases and which aren't, and it's platform-dependent. I actually implemented it like this at first in stringdtype and Marten suggested using a promoter, which was much simpler. See https://github.com/numpy/numpy-user-dtypes/blob/main/stringdtype/stringdtype/src/umath.c#L1362-L1385 for the approach with loops for each integer type.